_posts/2016-09-14-programming-model.md

```diff
@@ -0,0 +1,128 @@
+  - the allocated video memory storage for the texel data
+  - the way it's being accessed (format, binding)
+  - the sampler state to be used with it
+
+Dx11 introduced a conceptually new look at the graphics objects. It separated all these three concepts into different things. First, you create a texture object, which you can only have limited things to do with. If you want to use it as a render target, you create a render target view (RTV) for it. For sampling, you create a shader resource view (SRV), and may be an extra sampler object. There are also depth-stencil views (DSV) and unordered/ordered access views (UAV, OAV). You can have multiple different views into the same resource, as long as they are compatible (can't see a single-channel 16 bit format as RGBA8, for example).
+
+In GFX, when you create a texture you only provide the surface format, which dictates how many bits each channel has. Then, for the view creation, you specify how these bits are interpret (normalized integers, floats, raw ints, etc). Note: the resources are created by a `gfx::Factory` object, which may or may not be sendable/clonable, depending on the backend features. Example:
+{% highlight Rust %}
+let texture = try!(factory.create_texture(kind, levels, SHADER_RESOURCE | RENDER_TARGET, Usage::GpuOnly, None)); // providing the storage configuration and requesting the capabilities
+let resource = try!(factory.view_texture_as_shader_resource::<Srgba8>(&texture, (0, levels-1), format::Swizzle::new())); // creating an SRV for the whole mipmap range
+let target = try!(factory.view_texture_as_render_target(&texture, 0, None)); // creating an RTV for level 0
+{% endhighlight %}
+
+There are helper methods to create the texture with the views for your convenience (`create_render_target`, `create_depth_stencil`, etc). Later, the texture object itself does not participate in any rendering. It can only be used for updating the contents and reading them back. Typically, you'd ignore this object and only work with its views, given that they hold the parent object automatically.
+
+Note: applied to OpenGL, the GFX texture and RTV/DSV are mapping to the same thing (GL texture). The GL render buffer is used instead if `SHADER_RESOURCE` capability is not requested.
+
+
+## Pipeline states
+
+The general trend in graphics hardware is to do more and more with programmable shaders. This has been the case in consumer cards for a long time, since the first assembly shaders came into play. Thus, the states that used to map directly to switches in fixed-function hardware (like vertex attribute fetches) can now be compiled into a small chunk of shader code and executed somewhere in-between your shader code. This compile step and hookup can introduce lags and requires some anticipation by the driver, making those drivers rather bloated and full of quirks (part of the reason AMD/NV supply game profiles in the drivers). Current-gen APIs largely address this with pipeline states.
```

Comment:
> hus, the states that used to map directly to switches in fixed-function hardware (like vertex attribute fetches) can now be compiled into a small chunk of shader code and executed somewhere in-between your shader code. This compile step and hookup can introduce lags and requires some anticipation by the driver, making those drivers rather bloated and full of quirks

First mentions that PSO can now be compiled, following sentence describes the compilation step for PSO is reason for bloated drivers. Clarify that PSOs used to be handled by drivers and after mention that this is now handled by current-gen APIs?